db/commitlog/commitlog.cc

```diff
@@ -1132,7 +1132,12 @@ class db::commitlog::segment : public enable_shared_from_this<segment>, public c
             write(out, uint64_t(0));
         }
 
-        buf.remove_suffix(buf.size_bytes() - size);
+        auto to_remove = buf.size_bytes() - size;
+        // #20862 - we decrement usage counter based on buf.size() below.
+        // Since we are shrinking buffer here, we need to also decrement
+        // counter already
+        buf.remove_suffix(to_remove);
+        _segment_manager->totals.buffer_list_bytes -= to_remove;
```

Comment:
I expected to see an existing such substraction to be replaced by a new correct one. Previously, this buffer size was not substracted at all?

Reply:
buf.size() is subtracted after write is finished, down a few lines in the code. 
The reason I choose to subtract here already, instead of storing the "old" size and subtract it instead, later on, is that remove_suffix _might_ in fact release memory, hence it is somewhat appropriate  to already sub here.